bundles/org.eclipse.ui.ide.application/src/org/eclipse/ui/internal/ide/application/IDEApplication.java

```diff
@@ -332,17 +354,164 @@ protected Object checkInstanceLocation(Shell shell, Map applicationArguments) {
+		try (FileInputStream is = new FileInputStream(lockFile)) {
+			props.load(is);
+			String prop = props.getProperty(USER);
+			if (prop != null) {
+				sb.append("user: ").append(prop).append(System.lineSeparator()); //$NON-NLS-1$
+			}
+			prop = props.getProperty(HOST);
+			if (prop != null) {
+				sb.append("host: ").append(prop).append(System.lineSeparator()); //$NON-NLS-1$
+			}
+			prop = props.getProperty(DISPLAY);
+			if (prop != null) {
+				sb.append("display: ").append(prop).append(System.lineSeparator()); //$NON-NLS-1$
+			}
+			prop = props.getProperty(PROCESS_ID);
+			if (prop != null) {
+				sb.append("process-id: ").append(prop).append(System.lineSeparator()); //$NON-NLS-1$
+			}
+			return sb.toString();
+		} catch (Exception e) {
+			IDEWorkbenchPlugin.log("Could not read lock file: ", e); //$NON-NLS-1$
+		}
+		return null;
+	}
+
+	/**
+	 * Write lock owner details onto workspace lock file. Data includes user, host,
+	 * display and current java process id.
+	 */
+	private void writeWsLockInfo() {
+		Location instanceLoc = Platform.getInstanceLocation();
+		if (instanceLoc == null || instanceLoc.isReadOnly()) {
+			return;
+		}
+
+		File lockFile = getLockFile(instanceLoc.getURL());
+		if (lockFile == null) {
+			return;
+		}
+
+		Properties props = new Properties();
+
+		String user = System.getProperty(USER_NAME);
+		if (user != null) {
+			props.setProperty(USER, user);
+		}
+		String host = getHostName();
+		if (host != null) {
+			props.setProperty(HOST, host);
+		}
+		String display = getDisplay();
+		if (display != null) {
+			props.setProperty(DISPLAY, display);
+		}
+		String pid = getProcessId();
+		if (pid != null) {
+			props.setProperty(PROCESS_ID, pid);
+		}
+
+		try (OutputStream output = new FileOutputStream(lockFile)) {
+			props.store(output, null);
+		} catch (IOException e) {
+			IDEWorkbenchPlugin.log("Could not modify lock file", e); //$NON-NLS-1$
+		}
+	}
+
+	private String getDisplay() {
+		String displayEnv = null;
+		try {
+			displayEnv = System.getenv(DISPLAY_VAR);
+		} catch (Exception e) {
+			IDEWorkbenchPlugin.log("Failed to read DISPLAY variable.", e); //$NON-NLS-1$
+		}
+		return displayEnv;
+	}
+
+	private String getProcessId() {
+		Long pid = null;
+		try {
+			pid = ProcessHandle.current().pid();
+		} catch (Exception e) {
+			IDEWorkbenchPlugin.log("Failed to read Java process id.", e); //$NON-NLS-1$
+		}
+		return pid != null ? pid.toString() : null;
+	}
+
+	private String getHostName() {
+		String hostName = null;
+		try {
+			hostName = InetAddress.getLocalHost().getHostName();
+		} catch (UnknownHostException e) {
+			IDEWorkbenchPlugin.log("Failed to read host name.", e); //$NON-NLS-1$
+		}
+		return hostName;
+	}
+
+	private File getLockFile(URL workspaceUrl) {
+		if (workspaceUrl == null) {
+			return null;
+		}
+
+		// make sure the directory exists
+		File metaDir = new File(workspaceUrl.getPath(), METADATA_FOLDER);
+		if (!metaDir.exists()) {
```

Comment:
Please don't touch the directory! If the directory is missing, we have much bigger issue as missing lock info file and should **not** try to "fix" it in any way.

Reply:
For a new workspace lock directory is created by `org.eclipse.osgi.internal.location.BasicLocation.lock(File, URL)`.
As I am writing the data before we lock, So we need to create one.

Other two possible options I have are:
1. Initiate `org.eclipse.osgi.internal.location.BasicLocation.lock(File, URL)` twice.
If lock was successful in the first attempt, Then write the data`(This will unlock the file)`. Initiate a `one more lock again`.

2. Maintaining a `owner file` in along with lock file and read and write data there. This will not touch the lock file.

Reply:
**Please** read this comment: https://github.com/eclipse-platform/eclipse.platform.ui/pull/2347#discussion_r1791800009